### PR TITLE
Shopping: Floating Menu

### LINKED ISSUE #11252: Shopping: Floating Menu
## Feature Description

* [**Figma**](https://www.figma.com/file/ZrO763yvMWJjYeAPsRnNfW/_Editor---Sprint-Feature-Level-Designs_2022?node-id=503%3A15921)

Adding a dropdown to the floating menu for products that allows changing the displayed product

Screenshot 2022-04-12 at 14 41 23

Trash button: removes the element

### PR DESCRIPTION
**Pull request by timarney**
## Context

Add a floating menu with product dropdown for the selected product.

[Figma](https://www.figma.com/file/ZrO763yvMWJjYeAPsRnNfW/_Editor---Sprint-Feature-Level-Designs_2022?node-id=503%3A15921)

https://github.com/GoogleForCreators/web-stories-wp/issues/11252#issue-1201793703

## Summary

- Moves the current product dropdown into a component for re-use in the floating menu bar
- Adds floating menu bar
- Add trash button to remove the product

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Turn on the shopping feature (ensure you have some products available)
2. Add a story with a product 
3. Click on the product on the stage
4. Floating menu should appear with the product dropdown + trash button
5. Select a different product --- this should update the existing product
6. Delete the product using the `Trash` button

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11252

### CODE REVIEW

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import { useCallback, useEffect, useState } from '@googleforcreators/react';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { Datalist } from '@googleforcreators/design-system';
+import { useStory, useAPI } from '../../../../app';
+
+function ProductDropdown({ product, setProduct }) {
```
Comment (timarney) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js:38]: Most of this code was moved [from](https://github.com/GoogleForCreators/web-stories-wp/pull/11359/files#diff-26d7f3b93bff6c573c869429d2d19398803117f5a0d3c230231d6173ad2f29bfL51) into a standalone component so it can be used in the floating menu and also the panel.

File: packages/e2e-tests/src/specs/editor/shopping/productMenu.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * External dependencies
+ */
+import {
+  createNewStory,
+  withExperimentalFeatures,
+  minWPVersionRequired,
+  withPlugin,
+} from '@web-stories-wp/e2e-test-utils';
+
+describe('Shopping product', () => {
```
Comment (timarney) [inline packages/e2e-tests/src/specs/editor/shopping/productMenu.js:28]: Started a test --- this depends on https://github.com/GoogleForCreators/web-stories-wp/pull/11250 to get the products setup

File: packages/e2e-tests/src/specs/editor/shopping/testUtils.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { visitSettings } from '@web-stories-wp/e2e-test-utils';
+
+export const insertProduct = async (product) => {
```
Comment (timarney) [inline packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: Moved these from the `woocommerce` test --- Also updated those tests to use these utilities.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 4]
```diff
@@ -24,34 +23,13 @@ import {
...
   withPlugin,
-  visitSettings,
 } from '@web-stories-wp/e2e-test-utils';
 
 /**
  * Internal dependencies
  */
 import * as schema from './schema.json';
-
-async function insertProduct(product) {
```
Comment (timarney) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js:35]: Moved to testUtils.js

File: packages/story-editor/src/components/floatingMenu/menus/product.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    <>
+      <ProductDropdown
+        product={selectedElement?.product}
+        setProduct={setProduct}
+      />
+      <Separator />
+      <Trash
+        title={__('Remove product', 'web-stories')}
+        handleRemove={() => {
+          deleteElementById({ elementId: selectedElement.id });
```
Comment (swissspidy) [inline packages/story-editor/src/components/floatingMenu/menus/product.js]: There's a dedicated `deleteSelectedElements` reducer we can use here.

It does not make sense to render this dropdown for multi-selection, so we should avoid that. Then this reducer is safe to use.

File: packages/story-editor/src/components/floatingMenu/menus/product.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+          handleDismiss();
```
Comment (swissspidy) [inline packages/story-editor/src/components/floatingMenu/menus/product.js]: Is this needed? After deleting the currently selected element there ought to be no floating menu around anymore, no?

File: packages/story-editor/src/components/floatingMenu/menus/product.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    ({ actions, state }) => ({
+      updateElementById: actions.updateElementById,
+      selectedElement: state.selectedElements[0],
+      deleteElementById: actions.deleteElementById,
+    })
+  );
+
+  const setProduct = useCallback(
+    (product) => {
+      updateElementById({
```
Comment (swissspidy) [inline packages/story-editor/src/components/floatingMenu/menus/product.js]: Can use the `updateSelectedElements` reducer instead

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js (hunk 1 above)
Comment (swissspidy) [reply packages/story-editor/src/components/library/panes/shopping/productDropdown.js:38]: Once we refactor the panel then it's only going to be used here, right? At that point we can move it closer to the floating menu component I suppose.

File: packages/e2e-tests/src/specs/editor/shopping/testUtils.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  await expect(page).toClick('[aria-label="Product"]');
+  await page.waitForSelector('[aria-label="Search"]');
+
+  await page.type('[aria-label="Search"]', product);
+  await page.keyboard.press('ArrowDown');
+  await page.keyboard.press('Enter');
+  await expect(page).toClick('button', { text: 'Insert product' });
+};
+
+export const shoppingProvider = async (provider) => {
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: Nit: I prefer verbs in function names. What about something like `setShoppingProvider`?

**Review by swissspidy — COMMENTED** [review]
At first glance looks quite good, just a few minor notes.

This would be a great opportunity to add some Karma tests for the new floating menu items.

See #11314 

I know you haven't worked with these a lot so far, but especially Prometheus folks can help answer any questions. I'm around too.

Plus there are many existing Karma tests for the floating menu in other components already.

The biggest task is probably go add some fixture data for products to display in the dropdown. There's some precedence for that though in the code base.

PLMK if you have any questions.

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js (hunk 1 above)
Comment (timarney) [reply packages/story-editor/src/components/library/panes/shopping/productDropdown.js:38]: Correct it's not used in #11392 - can make adjustments after this one goes in.

File: packages/story-editor/src/components/floatingMenu/menus/product.js (hunk 6 above)
Comment (timarney) [reply packages/story-editor/src/components/floatingMenu/menus/product.js]: Seemed to be needed (but didn't re-test) as the code is now updated to use `deleteSelectedElements`

**Comment by timarney** [comment]
> This would be a great opportunity to add some Karma tests for the new floating menu items.
> 
> See #11314

I just self assigned myself to that issue and sounds good I'll dig into working with some Karma tests for this.

**Comment by spacedmonkey** [comment]
One thing that I noticed, was the the dropdown kind of jumps in a little. Not sure what we do about this, but it does look a little weird. 

https://user-images.githubusercontent.com/237508/166669277-a3c7d29a-747b-4567-80e1-c824d04667ac.mov

This jump maybe much more noticable, with shopify, if the API is slow or on slower hosting platforms, like shared hosts.

**Comment by timarney** [comment]
> One thing that I noticed, was the the dropdown kind of jumps in a little. Not sure what we do about this, but it does look a little weird.

Agreed I'll play around with that loading state a bit and see what I can do.

Screen Shot 2022-05-04 at 7 12 26 AM

**Comment by swissspidy** [comment]
Using styled-components:

```jsx
const StyledDatalist = styled(Datalist.DropDown)`
  width: 240px;
`;
```

**Comment by timarney** [comment]
@spacedmonkey --- that dropdown has a set width now so that helps as far as it not jumping.

File: packages/story-editor/src/components/floatingMenu/menus/product.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    useStory(({ actions, state }) => ({
+      updateSelectedElements: actions.updateSelectedElements,
+      selectedElement: state.selectedElements[0],
+      deleteSelectedElements: actions.deleteSelectedElements,
+    }));
+
+  const setProduct = useCallback(
+    (product) => {
+      updateSelectedElements({ properties: { product } });
+    },
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/floatingMenu/menus/product.js]:
```suggestion
    (product) => updateSelectedElements({ properties: { product } }),
```
Nitpic

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  useEffect(() => {
+    (async () => {
+      setIsLoading(true);
+      const products = await getProducts();
+      setInitialOptions(
+        products.map((p) => ({
+          name: p.productTitle,
+          id: p.productId,
+          product: p,
+        }))
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js]: This chunk of code is a repeat of logic in `getProductsByQuery`. Is there a chance to re-use this, even if it means submitting an empty string to `getProducts`

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+      );
+      setIsLoading(false);
+    })();
+  }, [getProducts]);
+
+  const onChange = ({ product: newProduct }) => {
+    setProduct(newProduct);
+  };
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js]:
```suggestion
  const onChange = ({ product: newProduct }) => setProduct(newProduct);
```

File: packages/e2e-tests/src/specs/editor/shopping/testUtils.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
+/*
```
Comment (spacedmonkey) [inline packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: Why is this file here? Why not in e2e-test-utils?

Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: I'll move it --- was looking at another directory which has it in `testUtils.js` not sure which one it was (doesn't matter). 

Screen Shot 2022-05-04 at 1 56 42 PM

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js (hunk 10 above)
Comment (timarney) [reply packages/story-editor/src/components/library/panes/shopping/productDropdown.js]: Updated here:
https://github.com/GoogleForCreators/web-stories-wp/pull/11359/files#diff-df9e7b1e3b42a1f29c8c10f8ecd4a1d1a7d1c80201a9d31f2c8f91bf02911683R57

File: packages/e2e-tests/src/specs/editor/shopping/testUtils.js (hunk 12 above)
Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: We usually only have testUtils folders for JS unit tests or Karma tests.

Comment (spacedmonkey) [reply packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: Why is this file still needed?

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        product: p,
+      }));
+    },
+    [getProducts]
+  );
+
+  useEffect(() => {
+    (async () => {
+      setIsLoading(true);
+      setInitialOptions(await getProductsByQuery());
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js]: If this await fails, then the loading state will not be reset. I think we should wrap this in a try / catch and put `setIsLoading(false)` in a finally.

File: packages/e2e-tests/src/specs/editor/shopping/testUtils.js (hunk 12 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/testUtils.js]: 🙃 forgot to remove -- just pushed an update

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+      setIsLoading(false);
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js]:
```suggestion
      try {
          setIsLoading(true);
          const products = await getProductsByQuery();
          setInitialOptions(products);
      } catch{
          setInitialOptions([]);
      } finally{
        setIsLoading(false);
      }
```

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js (hunk 13 above)
Comment (timarney) [reply packages/story-editor/src/components/library/panes/shopping/productDropdown.js]: Good call --- just pushed an update.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 15]
```diff
@@ -24,34 +23,13 @@ import {
...
-  // This clicks on the Product dropdown and waits for the dropdown menu to open.
-  await expect(page).toClick('[aria-label="Product"]');
-  await page.waitForSelector('[aria-label="Search"]');
-
-  await page.type('[aria-label="Search"]', product);
-  await page.keyboard.press('ArrowDown');
-  await page.keyboard.press('Enter');
-  await expect(page).toClick('button', { text: 'Insert product' });
-}
+import { insertProduct, setShoppingProvider } from './testUtils';
```
Comment (spacedmonkey) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Isn't this wrong now?

File: packages/story-editor/src/components/library/panes/shopping/productDropdown.js [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * Internal dependencies
+ */
+import { useStory, useAPI } from '../../../../app';
+
+const StyledDropDown = styled(Datalist.DropDown)`
+  width: 240px;
+`;
+function ProductDropdown({ product, setProduct }) {
+  const isEnabled = useFeature('shoppingIntegration');
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/shopping/productDropdown.js]: Maybe `isShoppingIntegrationEnabled` would be more clear

**Review by spacedmonkey — DISMISSED** [review]

File: packages/e2e-test-utils/src/shopping.js [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+/**
+ * Internal dependencies
+ */
+import visitSettings from './visitSettings';
```
Comment (swissspidy) [inline packages/e2e-test-utils/src/shopping.js]: Nit: 
```suggestion
/**
 * Internal dependencies
 */
import visitSettings from './visitSettings';
```

**Review by swissspidy — DISMISSED** [review]
Overall looks good, works well!

Thanks for setting up the Karma tests & fixture data as part of this!

**Review by swissspidy — APPROVED** [review]
